File: tests/python/test_community_ecology.py
"""
Tests for community-ecology-ordination skill scripts.
Run: pytest tests/python/test_community_ecology.py -v
"""
import pytest
import sys
import numpy as np
import pandas as pd
from pathlib import Path
from scipy.spatial.distance import braycurtis

DATA   = Path(__file__).parents[1] / "data"
SCRIPT = Path(__file__).parents[2] / "skills" / "community-ecology-ordination" / "scripts" / "community_analysis.py"

sys.path.insert(0, str(SCRIPT.parent))
import importlib.util
spec = importlib.util.spec_from_file_location("community_analysis", SCRIPT)
mod  = importlib.util.module_from_spec(spec)
spec.loader.exec_module(mod)


class TestAlphaDiversity:
    def make_sp(self, data):
        return pd.DataFrame(data, index=[f"s{i}" for i in range(len(data))])

    def test_richness_counts_nonzero(self):
        sp = self.make_sp([[5, 0, 3, 0], [0, 2, 0, 4], [1, 1, 1, 1]])
        div = mod.alpha_diversity(sp)
        assert list(div["richness"]) == [2, 2, 4]

    def test_shannon_zero_for_monoculture(self):
        sp = self.make_sp([[10, 0, 0]])
        div = mod.alpha_diversity(sp)
        assert div["shannon"].iloc[0] == pytest.approx(0.0, abs=1e-6)

    def test_shannon_max_for_equal_abundance(self):
        sp = self.make_sp([[5, 5, 5, 5]])
        div = mod.alpha_diversity(sp)
        expected_h = np.log(4)
        assert div["shannon"].iloc[0] == pytest.approx(expected_h, abs=0.01)

    def test_simpson_zero_for_monoculture(self):
        sp = self.make_sp([[100, 0, 0]])
        div = mod.alpha_diversity(sp)
        assert div["simpson"].iloc[0] == pytest.approx(0.0, abs=1e-6)

    def test_simpson_high_for_even_community(self):
        sp = self.make_sp([[25, 25, 25, 25]])
        div = mod.alpha_diversity(sp)
        assert div["simpson"].iloc[0] > 0.7


class TestBrayCurtis:
    def test_identical_sites_distance_zero(self):
        a = np.array([5.0, 3.0, 0.0, 2.0])
        b = np.array([5.0, 3.0, 0.0, 2.0])
        assert braycurtis(a, b) == pytest.approx(0.0, abs=1e-10)

    def test_non_overlapping_sites_distance_one(self):
        a = np.array([5.0, 0.0, 0.0])
        b = np.array([0.0, 3.0, 2.0])
        assert braycurtis(a, b) == pytest.approx(1.0, abs=1e-10)

    def test_matrix_is_symmetric(self):
        sp = pd.read_csv(DATA / "species_site_matrix.csv", index_col=0).drop(columns=["group"])
        dm = mod.bray_curtis_matrix(sp)
        assert np.allclose(dm, dm.T, atol=1e-10)

    def test_diagonal_is_zero(self):
        sp = pd.read_csv(DATA / "species_site_matrix.csv", index_col=0).drop(columns=["group"])
        dm = mod.bray_curtis_matrix(sp)
        assert np.allclose(np.diag(dm), 0.0, atol=1e-10)

    def test_values_in_range(self):
        sp = pd.read_csv(DATA / "species_site_matrix.csv", index_col=0).drop(columns=["group"])
        dm = mod.bray_curtis_matrix(sp)
        assert dm.min() >= 0.0
        assert dm.max() <= 1.0 + 1e-10


class TestGroupSeparation:
    def test_within_group_lower_than_between(self):
        """Forest sites should be more similar to each other than to pasture sites."""
        sp = pd.read_csv(DATA / "species_site_matrix.csv", index_col=0)
        groups = sp["group"]
        sp = sp.drop(columns=["group"])
        dm = mod.bray_curtis_matrix(sp)
        idx_forest  = [i for i, g in enumerate(groups) if g == "forest"]
        idx_pasture = [i for i, g in enumerate(groups) if g == "pasture"]
        within  = np.mean([dm[i,j] for i in idx_forest  for j in idx_forest  if i<j])
        between = np.mean([dm[i,j] for i in idx_forest  for j in idx_pasture])
        assert within < between, f"Within-forest ({within:.3f}) should be < between ({between:.3f})"


class TestEndToEnd:
    def test_script_produces_outputs(self, tmp_path):
        import subprocess, sys
        result = subprocess.run(
            [sys.executable, str(SCRIPT),
             str(DATA / "species_site_matrix.csv"),
             str(DATA / "site_metadata.csv"),
             str(tmp_path)],
            capture_output=True, text=True
        )
        assert result.returncode == 0, f"Script failed:\n{result.stderr}"
        assert (tmp_path / "diversity_metrics.csv").exists()
        assert (tmp_path / "bray_curtis_matrix.csv").exists()
        div = pd.read_csv(tmp_path / "diversity_metrics.csv")
        assert len(div) == 30
        assert (div["richness"] > 0).all()

File: tests/r/test-community-ecology.R
# ecological-agent-skills / Copyright (C) 2026 Francisco Diego Barros Barata
# SPDX-License-Identifier: GPL-3.0-or-later

# test-community-ecology.R
library(testthat)

DATA_DIR <- if (dir.exists("tests/data")) "tests/data" else file.path("..", "data")

bray_curtis <- function(a, b) {
  sum(abs(a - b)) / (sum(a) + sum(b))
}

alpha_diversity <- function(row) {
  row <- row[row > 0]
  if (length(row) == 0) return(c(richness=0, shannon=0, simpson=0))
  p <- row / sum(row)
  list(richness = length(row),
       shannon  = -sum(p * log(p)),
       simpson  = 1 - sum(p^2))
}

test_that("species matrix loads correctly", {
  df <- read.csv(file.path(DATA_DIR, "species_site_matrix.csv"), row.names = 1)
  expect_equal(nrow(df), 30)
  expect_true("group" %in% names(df))
})

test_that("three groups are balanced", {
  df <- read.csv(file.path(DATA_DIR, "species_site_matrix.csv"), row.names = 1,
                 stringsAsFactors = FALSE)
  counts <- table(df$group)
  expect_equal(as.integer(counts["forest"]),  10)
  expect_equal(as.integer(counts["savanna"]), 10)
  expect_equal(as.integer(counts["pasture"]), 10)
})

test_that("no negative abundances", {
  df <- read.csv(file.path(DATA_DIR, "species_site_matrix.csv"), row.names = 1)
  sp <- df[, !names(df) %in% "group"]
  expect_true(all(sp >= 0))
})

test_that("Bray-Curtis is 0 for identical sites", {
  a <- c(5, 3, 0, 2)
  b <- c(5, 3, 0, 2)
  expect_equal(bray_curtis(a, b), 0.0, tolerance = 1e-10)
})

test_that("Bray-Curtis is 1 for non-overlapping sites", {
  a <- c(5, 0, 0)
  b <- c(0, 3, 2)
  expect_equal(bray_curtis(a, b), 1.0, tolerance = 1e-10)
})

test_that("Shannon is 0 for monoculture", {
  div <- alpha_diversity(c(100, 0, 0))
  expect_equal(div$shannon, 0.0, tolerance = 1e-10)
})

test_that("Shannon is log(n) for perfectly even community", {
  div <- alpha_diversity(c(5, 5, 5, 5))
  expect_equal(div$shannon, log(4), tolerance = 0.01)
})

test_that("vegan NMDS runs on test data", {
  skip_if_not_installed("vegan")
  df <- read.csv(file.path(DATA_DIR, "species_site_matrix.csv"), row.names = 1)
  sp <- df[, !names(df) %in% "group"]
  set.seed(42)
  nmds <- vegan::metaMDS(sp, distance = "bray", k = 2, trymax = 20, trace = FALSE)
  expect_lt(nmds$stress, 0.25)
})

test_that("forest sites more similar to each other than to pasture", {
  skip_if_not_installed("vegan")
  df     <- read.csv(file.path(DATA_DIR, "species_site_matrix.csv"), row.names = 1)
  groups <- df$group
  sp     <- df[, !names(df) %in% "group"]
  dm     <- as.matrix(vegan::vegdist(sp, method = "bray"))
  ff_idx <- which(groups == "forest")
  pp_idx <- which(groups == "pasture")
  within  <- mean(dm[ff_idx, ff_idx][lower.tri(dm[ff_idx, ff_idx])])
  between <- mean(dm[ff_idx, pp_idx])
  expect_lt(within, between)
})

test_that("PERMANOVA detects group effect", {
  skip_if_not_installed("vegan")
  df     <- read.csv(file.path(DATA_DIR, "species_site_matrix.csv"), row.names = 1,
                     stringsAsFactors = FALSE)
  groups <- df$group
  sp     <- df[, !names(df) %in% "group"]
  set.seed(42)
  result <- vegan::adonis2(sp ~ groups, data = data.frame(groups), method = "bray",
                           permutations = 199)
  expect_lt(result$`Pr(>F)`[1], 0.05)
})
